### PR TITLE
Setting up  dagger

### PR DESCRIPTION
**Pull request by blazsolar**
This is first step into Dagger introduction to this project as to #13 issue. This would be first of many pull requests required to implement dagger across the app. Maybe it would be good to create a separate branch to do full dagger implementation in it?

### CODE REVIEW

File: base/src/main/java/butter/droid/base/data/DataModule.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Butter. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+ * This file is part of Butter.
```
Comment (se-bastiaan) [inline base/src/main/java/butter/droid/base/data/DataModule.java]: Double copyright notice

**Comment by se-bastiaan** [comment]
If you have the time for this then you're welcome to fully implement Dagger in the current codebase, I don't see any changes coming from my side in the coming weeks. Then I would be able to squash the commits and make it one commit on develop.

**Comment by xaiki** [comment]
is this still being worked on ?

**Comment by blazsolar** [comment]
yes it is, but the progress is slow due to lack of time.

**Comment by xaiki** [comment]
great to see advance on this ! please ping us up when you feel it's ready for merge, as android is pretty much unmaintained as soon as we can build this, we can put it in ! thanks !

**Comment by blazsolar** [comment]
i still need to go through everything and make sure i didn't miss anything. Also CI has to be fixed. Also one of you guys should probably review it.

**Comment by xaiki** [comment]
maybe @cannnon can review plez pretty plez ?

File: base/src/main/java/butter/droid/base/manager/provider/ProviderManager.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+    @ProviderType private int currentProviderType;
+
+    private final List<OnProviderChangeListener> listeners = new ArrayList<>();
+
+    public ProviderManager(@Nullable MediaProvider movieProvider, @Nullable MediaProvider showProvider,
+            @Nullable SubsProvider subsProvider) {
+        this.movieProvider = movieProvider;
+        this.showProvider = showProvider;
+        this.subsProvider = subsProvider;
```
Comment (MrRory) [inline base/src/main/java/butter/droid/base/manager/provider/ProviderManager.java]: This way only one `SubsProvider` will be able to be used, but it should be able to differ based on the actual `MediaProvider`.

**Comment by MrRory** [comment]
maybe I can? i placed one comment about the subsprovider but it actually is relevant as a whole. the way the mediaproviders are provided doesnt allow for any flexibility. there is only a movie provider, a tv provider and _one_ subsprovider. what if we'd want to add another provider?

**Comment by xaiki** [comment]
thanks for your review @MrRory. right now we already have hardcoded providers in the android app. this PR is only to integrate dagger, for the multiple provider work, we should get another one.

File: base/src/main/java/butter/droid/base/manager/provider/ProviderManager.java (hunk 2 above)
Comment (blazsolar) [inline base/src/main/java/butter/droid/base/manager/provider/ProviderManager.java]: Ok, so every media type can have its own subs provider, if i understand correctly?

**Comment by blazsolar** [comment]
I did a quick change so that subs provider is tied to media provider as mentioned above. As far as I am concerned this is now it for basic dagger implementation and will be a good base to work on from. CI fixes will be done new PR.

**Comment by xaiki** [comment]
i don't know much about java nor android, but that build is failing

**Comment by vzamanillo** [comment]
The build fail is related to the gradle memory, we need more memory to complete all tasks, we need to uncoment the [memory settings](https://github.com/popcorn-official/popcorn-android/blob/development/gradle.properties#L13) in the gradle properties to get a succesful build.

**Comment by blazsolar** [comment]
won't work. Container on travis has only 4GB of memory available to you. Every gradle module uses specified amount (max). And we are also running emulator which takes some memory. 

I'll try using `sudo: required` option that uses the old non container architecture and offers more memory. I think it is 7.9GB

**Comment by blazsolar** [comment]
travis is now working as is should, therefore this branch is ready to merge

**Comment by vzamanillo** [comment]
👍

**Comment by xaiki** [comment]
thanks all for the hard work on getting this in !

**Pull request merged**